examples/node/socket.c

```diff
@@ -0,0 +1,302 @@
+    }
+
+    struct addrinfo* res = NULL;
+    long port;
+    char* endptr;
+
+    int i;
+    for (i = addr_len - 1; i >= 0; i--)
+    {
+        if (addr_buf[i] == ':') break;
+    }
+
+    if (addr_buf[i] != ':')
+    {
+        NODE_ERROR("Malformed address:port string: '%s'", addr);
+        goto end;
+    }
+
+    addr_buf[i] = '\0';
+    port = strtol(addr_buf + i + 1, &endptr, 10);
+
+    if (port <= 0 || port > USHRT_MAX || errno ||
+        (*endptr != '\0' && !isspace(*endptr)))
+    {
+        NODE_ERROR("Malformed/invalid port: '%s'. Errno: %d (%s)",
+                   addr_buf + i + 1, errno, strerror(errno));
+        goto end;
+    }
+
+    res = socket_get_addrinfo2(strlen(addr_buf) > 0 ? addr_buf : NULL,
+                               (uint16_t)port);
+end:
+    free(addr_buf);
+    return res;
+}
+
+static struct node_socket*
+socket_create(int const fd)
+{
+    assert(fd > 0);
+
+    struct node_socket* res = calloc(1, sizeof(struct node_socket));
+    if (res)
+    {
+        res->fd = fd;
+    }
+    else
+    {
+        NODE_ERROR("Failed to allocate struct node_socket: %d (%s)",
+                   errno, strerror(errno));
+        close(fd);
+    }
+
+    return res;
+}
+
+/**
+ * Definition of function type with the signature of bind() and connect()
+ */
+typedef int (*socket_act_fun_t) (int                    sock_fd,
+                                 const struct sockaddr* addr,
+                                 socklen_t              addrlen);
+
+static int
+socket_bind_and_listen(int                    const sfd,
+                       const struct sockaddr* const addr,
+                       socklen_t              const addrlen)
+{
+    int ret = bind(sfd, addr, addrlen);
+
+    if (!ret)
+        ret = listen(sfd, SOMAXCONN);
+
+    return ret;
+}
+
+/**
+ * A "template" method to do the "right thing" with the addrinfo and create a
+ * socket from it.
+ *
+ * @param[in] info       addrinfo list, swallowed and deallocated
+ * @param[in] action_fun the "right thing" to do on socket and struct sockaddr
+ * @param[in] action_str action description to be printed in the error message
+ * @param[in] orig_host  host address to be pronted in the error message
+ * @param[in] orig_port  port to be printed in the error message, if orig_host
+ *                       string contains the port, this parameter should be 0
+ *
+ * The last three parameters are for diagnostic puposes only. orig_host and
+ * orig_port are supposed to be what were used to obtain addrinfo.
+ *
+ * @return new struct node_socket.
+ */
+static struct node_socket*
+socket_from_addrinfo(struct addrinfo* const info,
+                     socket_act_fun_t const action_fun,
+                     const char*      const action_str,
+                     const char*      const orig_host,
+                     uint16_t         const orig_port)
+{
+    int sfd;
+    int err = 0;
+
+    /* iterate over addrinfo list and try to bind a listening socket */
```

Comment:
This comment is a bit misleading as this function seems to be used for both listening and connecting.

Reply:
good catch! It's all my code deduplicaton efforts. Thanks!